crates/wasi-nn/src/backend/pytorch.rs

```diff
@@ -0,0 +1,174 @@
+        )?;
+
+        // Set the model to be used for inference (eval), default mode is training.
+        compiled_module.f_set_eval()?;
+
+        let graph = PytorchGraph(Arc::new(Mutex::new(compiled_module)));
+        let box_: Box<dyn BackendGraph> = Box::new(graph);
+        Ok(box_.into())
+    }
+
+    fn as_dir_loadable<'a>(&'a mut self) -> Option<&'a mut dyn BackendFromDir> {
+        Some(self)
+    }
+}
+
+impl BackendFromDir for PytorchBackend {
+    fn load_from_dir(
+        &mut self,
+        path: &Path,
+        target: ExecutionTarget,
+    ) -> Result<Graph, BackendError> {
+        // Load the model from the file path
+        let compiled_module =
+            CModule::load_on_device(path, map_execution_target_to_string(target)).unwrap();
+        let graph = PytorchGraph(Arc::new(Mutex::new(compiled_module)));
+        let box_: Box<dyn BackendGraph> = Box::new(graph);
+        Ok(box_.into())
+    }
+}
+
+struct PytorchGraph(Arc<Mutex<tch::CModule>>);
+
+unsafe impl Send for PytorchGraph {}
+unsafe impl Sync for PytorchGraph {}
+
+impl BackendGraph for PytorchGraph {
+    fn init_execution_context(&self) -> Result<ExecutionContext, BackendError> {
+        let tensor = TchTensor::new();
+        let box_: Box<dyn BackendExecutionContext> =
+            Box::new(PytorchExecutionContext(self.0.clone(), tensor));
+        Ok(box_.into())
+    }
+}
+
+unsafe impl Sync for PytorchExecutionContext {}
+struct PytorchExecutionContext(Arc<Mutex<tch::CModule>>, tch::Tensor);
+
+impl BackendExecutionContext for PytorchExecutionContext {
+    fn set_input(&mut self, _index: Id, input_tensor: &Tensor) -> Result<(), BackendError> {
+        // Input index is not used in pytorch models. The forward method to a model passes the tensor/data to
+        // the appropriate layer of the model.
+        let kind = input_tensor.ty.try_into()?;
+        let dimensions = input_tensor
+            .dimensions
+            .iter()
+            .map(|&dim| dim as i64)
+            .collect::<Vec<_>>();
+        self.1 = TchTensor::from_data_size(&input_tensor.data, &dimensions, kind);
+        Ok(())
+    }
+
+    fn compute(&mut self) -> Result<(), BackendError> {
+        // Use forward method on the compiled module/model after locking the mutex, and pass the input tensor to it
+        self.1 = self.0.lock().unwrap().forward_ts(&[&self.1]).unwrap();
+        Ok(())
+    }
+
+    fn get_output(&mut self, _index: Id) -> Result<Tensor, BackendError> {
+        // Output index is not used in pytorch models. The forward method to a model returns the output tensor.
+        let numel = self.1.numel();
+        let dimensions = self.1.size();
+        let ty = self.1.kind().try_into()?;
+        let data = vec![0f32; numel];
+        let mut data_u8: Vec<u8> = data
+            .iter()
+            .flat_map(|&x| x.to_le_bytes().to_vec())
+            .collect();
+        self.1.copy_data_u8(&mut data_u8, numel);
```

Comment:
This is incorrect: we need to retrieve the data regardless of the type, so we need to first figure out how many bytes each `Kind` is before constructing the receiving buffer, like:

```suggestion
        let data = vec![0u8; size_of(ty) * numel];
        self.1.copy_data_u8(&mut data, numel);
```

Reply:
Done, added `kind_to_size()` for conversions.